User-CRUD

meb-wtag opened the pull request:
[JTJ-6 User-CRUD](https://welltravel.atlassian.net/jira/software/projects/JTJ/boards/44?selectedIssue=JTJ-6)

https://github.com/meb-wtag/journey_me/assets/141820874/4991b880-7738-4897-9c5d-269ec085e04a

Created User_CRUD including:
- Login Page (view)
- Register Page (view)
- Edit User Page (view)
- Show User Page (view)
- created UserSession Controller
- updated UserController
- Active Storage
- Feature to upload a Profile Picture (including JS)
- Tests for all Controller and Models
- Journals are now dependent on Users
- (before_action: :require_login) Disable access on foreign pages

Extra Views and Features
- Added Sidebar to navigate
- Implemented I18n

meb-wtag committed:
.

meb-wtag committed:
Merge branch 'master' into User-CRUD

meb-wtag committed:
CRUD for User

tgi-wtag commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,52 @@
+  end
+
+  def edit
+    @user = User.find(params[:id]) # Replace with your logic to fetch the user record
```
?

meb-wtag committed:
Format changes

tgi-wtag commented on Gemfile:
```diff
@@ -15,14 +15,16 @@ gem "tzinfo-data", platforms: %i[ mingw mswin x64_mingw jruby ]
 gem 'bootstrap', '~> 5.3.1'
 gem 'simple_form'
+gem 'country_select'
+
```
delete lines

tgi-wtag commented on app/models/user.rb:
```diff
@@ -0,0 +1,7 @@
+class User < ApplicationRecord
+
+	enum role: { admin: 0, superuser: 1, user: 2}
+
```
delete line

tgi-wtag commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,52 @@
+  def destroy
+    @user = User.find(params[:id])
+    @user.destroy
+
```
delete line

tgi-wtag commented on app/views/users/new.html.erb:
```diff
@@ -0,0 +1,28 @@
+  <%= f.input :last_name %>
+  <%= f.input :country %>
+  <%= f.collection_radio_buttons :role, 
+  [['Admin', 0], ['Superuser', 1], ['User', 2]], 
```
indent

meb-wtag replied:
in my opinion not necessary :)

tgi-wtag commented on app/views/users/new.html.erb:
```diff
@@ -0,0 +1,28 @@
+  :second, :first %>
+  <%= f.input :description %>
+  <%= f.input :date_of_birth%>
+
```
delete line

tgi-wtag commented on app/views/users/new.html.erb:
```diff
@@ -0,0 +1,28 @@
+  <%= f.input :date_of_birth%>
+
+  </div>
+
```
delete line

tgi-wtag commented on app/views/users/new.html.erb:
```diff
@@ -0,0 +1,28 @@
+  <div class="btn btn-primary">
+    <%= f.button :submit, {class: "form-check"} %>
+  </div>
+
```
delete line

tgi-wtag commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,113 @@
+            <%= link_to "Delete", user_path(@user), class: "btn btn-danger", data: {
+                    turbo_method: :delete,
+                    turbo_confirm: "Are you sure?"
+                  } %>
```
?

tgi-wtag commented on db/migrate/20230904071759_create_users.rb:
```diff
@@ -0,0 +1,27 @@
+      t.string :last_name
+      t.integer :role, default: 2
+      t.text :description
+
```
delete line

lra-wtag replied:
these empty lines are not needed

tgi-wtag commented on db/migrate/20230904071759_create_users.rb:
```diff
@@ -0,0 +1,27 @@
+
+      t.string :email
+      t.integer :mobile
+
```
delete line

lra-wtag replied:
these empty lines are not needed

tgi-wtag commented on db/migrate/20230904071759_create_users.rb:
```diff
@@ -0,0 +1,27 @@
+      t.string :adress
+      t.string :country
+      t.string :city
+      
```
delete line

lra-wtag replied:
these empty lines are not needed

tgi-wtag commented on db/migrate/20230904071759_create_users.rb:
```diff
@@ -0,0 +1,27 @@
+      
+      t.date :date_of_birth 
+      t.date :joining_date
+
```
delete line

lra-wtag replied:
these empty lines are not needed

tgi-wtag commented on test/models/user_test.rb:
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+class UserTest < ActiveSupport::TestCase
+  # test "the truth" do
```
remove comment

meb-wtag committed:
Fixxed PR comments

All  known Problems from PR resolved

lra-wtag commented on app/models/user.rb:
```diff
@@ -0,0 +1,7 @@
+class User < ApplicationRecord
+
+	enum role: { admin: 0, superuser: 1, user: 2}
```
```suggestion
	enum role: { admin: 0, super_user: 1, user: 2}
```
whats a super user?

meb-wtag replied:
supervisor

lra-wtag commented on test/fixtures/users.yml:
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
delete this file

lra-wtag commented on test/models/user_test.rb:
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
```
delete this file

meb-wtag committed:
PR changes #1

Deleted one file and changed format of 3 others.

lra-wtag commented on app/models/user.rb:
```diff
@@ -0,0 +1,6 @@
+	enum role: { admin: 0, supervisor: 1, user: 2}
+
+	validates :first_name, :last_name, presence: true
+end
```
please fix the indentation of this file

lra-wtag commented on app/views/users/new.html.erb:
```diff
@@ -0,0 +1,27 @@
+  :second, :first %>
+  <%= f.input :description %>
+  <%= f.input :date_of_birth%>
+  </div>
```
```
<div class="form-inputs">
    <%= f.input :username %>
    <%= f.input :first_name %>
    <%= f.input :last_name %>
    <%= f.input :country %>
    <%= f.collection_radio_buttons :role, [['Admin', 0], ['Superuser', 1], ['User', 2]], :second, :first %>
    <%= f.input :description %>
    <%= f.input :date_of_birth%>
</div>
```

meb-wtag committed:
Update user.rb

meb-wtag committed:
Changes in format

lra-wtag reviewed (changes requested):
Lets start writing the model and controller specs from here

meb-wtag committed:
Update Gemfile.lock

meb-wtag committed:
Creating Testing environment

meb-wtag committed:
Creating the test for Controller

meb-wtag committed:
adjusting tests

meb-wtag committed:
Creating all the tests without Index

meb-wtag committed:
.

meb-wtag committed:
Created the controller SPECS

meb-wtag committed:
Created Model Specs

meb-wtag committed:
Update users_controller_spec.rb

meb-wtag committed:
Updating tests

meb-wtag committed:
implementing i18n

meb-wtag committed:
Implemented i18n

evk-wtag commented on db/migrate/20230904071759_create_users.rb:
```diff
@@ -0,0 +1,24 @@
+class CreateUsers < ActiveRecord::Migration[7.0]
+  def up
+      create_table :users do |t|
```
Indentation

evk-wtag commented on db/migrate/20230904071759_create_users.rb:
```diff
@@ -0,0 +1,24 @@
+  def up
+      create_table :users do |t|
+      t.string :username
+      t.string :first_name
+      t.string :last_name
```
Let's add the db level validations too

lra-wtag commented on app/helpers/user_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module UserHelper
```
delete this file if not needed

lra-wtag commented on spec/helpers/user_helper_spec.rb:
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
```
delete this file if not needed

lra-wtag commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,12 @@
+    user = User.new(first_name: 'Name')
+    expect(user.valid?).to be_falsey
+  end 
+end
```
fix the indentation of this file

lra-wtag replied:
Or you can also write it as below:
```
it { is_expected.to validate_presence_of(:first_name) }
it { is_expected.to validate_presence_of(:last_name) }
```

Also I am missing the enum values specs here

meb-wtag committed:
Update schema.rb

meb-wtag committed:
Update schema.rb

meb-wtag committed:
Merge remote-tracking branch 'origin/master' into User-CRUD

meb-wtag committed:
SEssion CRud

meb-wtag committed:
Sidebar/Login Page

Adding Sidebar to navigate

meb-wtag committed:
Created Login Functions

-View for Login
-Login controller

evk-wtag commented on app/assets/stylesheets/application.css:
```diff
@@ -1,15 +1,11 @@
+
+.div1 { grid-area: 1 / 2 / 6 / 5; }
+.div2 { grid-area: 1 / 1 / 6 / 2; }
+.div3 { grid-area: 1 / 5 / 6 / 6; }
```
Please add a new line at EOF

evk-wtag commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,14 @@
 class ApplicationController < ActionController::Base
+	helper_method :current_user
```
```suggestion
  helper_method :current_user
```

evk-wtag commented on app/controllers/user_sessions_controller.rb:
```diff
@@ -0,0 +1,23 @@
+      flash[:alert] = "Success"
+      redirect_to user_path(current_user)
+    else
+      flash[:alert] = "Login failed"
```
Let's use i18n for this

evk-wtag commented on app/controllers/user_sessions_controller.rb:
```diff
@@ -0,0 +1,23 @@
+  end
+end
+
+def current_user
+  @current_user ||= User.find(params[:id])
+end
```
This method is outside the class...

evk-wtag commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,59 @@
+class UsersController < ApplicationController
+   before_action :find_user, only: %i[show destroy update]
```
```suggestion
  before_action :find_user, only: %i[show destroy update]
```

evk-wtag commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,59 @@
+    end
+  end
+
+  def show;end
```
Let's remove if not needed

evk-wtag commented on app/helpers/user_sessions_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module UserSessionsHelper
```
Let's remove this file if not needed

evk-wtag commented on app/views/layouts/application.html.erb:
```diff
@@ -15,15 +15,18 @@
-    <%= yield %>
+  <% flash.each do |type, message| %>
+    <div class="alert alert-<%= type %> text-center">
+      <%= message %>
```
Missing i18n

evk-wtag commented on app/views/user_sessions/new.html.erb:
```diff
@@ -0,0 +1,14 @@
+  <p>
+    <%= f.submit 'Login' %>
+  </p>
+<% end %>
```
New line at EOF for all files please!

evk-wtag commented on config/routes.rb:
```diff
@@ -1,5 +1,7 @@
 Rails.application.routes.draw do
   root 'journals#index'
+  resources :user_sessions do; end
```
```suggestion
  resources :user_sessions 
```
 and so on..

meb-wtag committed:
Created Login

-Methods for: Logout, Login, Register
- Added pages to all of these

Updated Method Edit User

meb-wtag committed:
Corrected all the logIn views

meb-wtag committed:
Creating the tests

meb-wtag committed:
Update show.html.erb

meb-wtag committed:
Update show.html.erb

meb-wtag committed:
Modified Authentication

Added Feature so that no user can access any other pages then his own ones. Futhermore i implemented a feature to upload a PP and gave it some fancy animations

meb-wtag committed:
,

meb-wtag committed:
Fixed smaller issues befor creating the PR

meb-wtag committed:
Final RuboCop Correct

At the end I checked and Autocorrected all the files with RuboCop.

meb-wtag committed:
Revert "Final RuboCop Correct"

This reverts commit a26ab3803660014f549856df859889a84513a552.

evk-wtag commented on app/assets/config/manifest.js:
```diff
@@ -1,4 +1,5 @@
+//= link application.js 
 //= link_tree ../../javascript .js
-//= link_tree ../../../vendor/javascript .js
+//= link_tree ../../../vendor/javascript .js
```
New line at EOF please

evk-wtag commented on app/assets/stylesheets/application.css:
```diff
@@ -1,15 +1,143 @@
- *= require_self
- */
+.parent {
+display: grid;
+grid-template-columns: repeat(5, 1fr);
+grid-template-rows: repeat(5, 1fr);
+grid-column-gap: 0px;
+grid-row-gap: 0px;
```
Not indented

evk-wtag commented on app/controllers/.DS_Store:
Let's remove this file

evk-wtag commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,17 @@
+  def require_login
+    user_id = params[:id].to_i
+    unless current_user == @user
+      flash[:error] = "You must be logged in to access this page."
```
i18n for this please

evk-wtag commented on config/routes.rb:
```diff
@@ -1,7 +1,13 @@
+  delete 'user_sessions', to: 'user_sessions#destroy', as: :logout
+
+  resources :users do
+    post 'upload_profile_picture', to: 'users#upload_profile_picture', on: :collection
```
```suggestion
    post :upload_profile_picture, to: 'users#upload_profile_picture', on: :collection
```

evk-wtag commented on config/routes.rb:
```diff
@@ -1,7 +1,13 @@
-    resources :journal_entries do
+  root 'users#new'
+  resources :user_sessions, only: [:new, :create] 
+  delete 'user_sessions', to: 'user_sessions#destroy', as: :logout
```
```suggestion
  delete :user_sessions, to: 'user_sessions#destroy', as: :logout
```

evk-wtag commented on Gemfile:
```diff
@@ -3,13 +3,18 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 gem 'bootsnap', require: false
 gem 'bootstrap', '~> 5.3.1'
+gem 'bcrypt', '~> 3.1.7'
+gem 'cancancan'
```
Are we implementing cancancan abilities in this PR? If not, lets remove this gem. Let's add it back when we are introducing the functionalities.

evk-wtag commented on db/migrate/20230904071759_create_users.rb:
```diff
@@ -0,0 +1,26 @@
+      t.string :username, null: false, uniqueness: true
+      t.string :first_name
+      t.string :last_name
+      t.integer :role, default: 2
```
What are the roles? Wouldn't we add it in model?

evk-wtag commented on app/assets/stylesheets/application.css:
Can we have multiple css files for the different components? Let's not add all the css in one file. Let's breakdown the css into separate components and use this file to import the individual css files

meb-wtag committed:
Correction 1

Corrections on format

meb-wtag committed:
RuboCop on all edited files

evk-wtag commented on app/assets/stylesheets/grid.css:
```diff
@@ -0,0 +1,11 @@
+.parent {
+	display: grid;
+	grid-template-columns: repeat(5, 1fr);
+	grid-template-rows: repeat(5, 1fr);
+	grid-column-gap: 0px;
+	grid-row-gap: 0px;
```
indentation

meb-wtag replied:
what's incorrect in here?

evk-wtag commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,17 @@
+  end
+
+  def require_login
+    params[:id].to_i
```
how is this working?

meb-wtag replied:
fortgot to delete this line

evk-wtag commented on app/models/ability.rb:
Do we need this file now? I think you removed the gem

evk-wtag commented on app/views/layouts/_sidebar.html.erb:
```diff
@@ -0,0 +1,24 @@
+      			<%= link_to( t('sidebar.journals'), user_journals_path(current_user), {class:"nav-link link-dark"} )%><br> 
+      </li>
+      <li>
+            <%= link_to( t('sidebar.new_user'), new_user_path, {class:"nav-link link-dark"} )%><br> 
```
Indentation

evk-wtag commented on app/views/layouts/_sidebar.html.erb:
```diff
@@ -0,0 +1,24 @@
+<div class="d-flex flex-column flex-shrink-0 p-3 bg-light" style="">
+    <a href="/" class="d-flex align-items-center mb-3 mb-md-0 me-md-auto link-dark text-decoration-none">
+      <span class="fs-4">Journey</span>
```
i18n please

evk-wtag commented on app/views/users/index.html.erb:
```diff
@@ -0,0 +1,6 @@
+<% else %>
+  <h1>This is the index page</h1>
+  <%= link_to t('user_session.action.add'), new_user_path %><br>
+<% end %>
```
Missing i18n and new line at end of file

evk-wtag commented on config/locales/en.yml:
```diff
@@ -40,3 +40,46 @@ en:
+    journals: Journals
+    new_user: New User
+    login: Login
+    journey: Journey    
```
Please add new line at EOF. Please check all files

evk-wtag commented on db/schema.rb:
```diff
@@ -29,6 +57,28 @@
+    t.string "username", null: false
+    t.string "first_name"
+    t.string "last_name"
+    t.integer "role", default: 2
```
I guess you removed roles from the migration but it still exists in the schema file. Please run the migrations again and keep this up to date...

evk-wtag commented on Gemfile.lock:
```diff
@@ -87,6 +90,7 @@ GEM
       popper_js (>= 2.11.8, < 3)
       sassc-rails (>= 2.0.0)
     builder (3.2.4)
+    cancancan (3.5.0)
```
Bundle install again please 😅

evk-wtag commented on spec/helpers/user_sessions_helper_spec.rb:
```diff
@@ -0,0 +1,15 @@
+#       expect(helper.concat_strings("this","that")).to eq("this that")
+#     end
+#   end
+# end
```
Let's remove the comments if not required

evk-wtag commented on spec/requests/user_sessions_spec.rb:
Please remove this file if it's not used for proper testing

evk-wtag commented on db/migrate/20230831073432_create_journals.rb:
```diff
@@ -1,6 +1,6 @@
 class CreateJournals < ActiveRecord::Migration[7.0]
   def up
-    create_table :journals do |t|
+    create_table :journals do |t| 
```
```suggestion
    create_table :journals do |t|
```

meb-wtag committed:
PR correction 2

evk-wtag commented on app/assets/stylesheets/grid.css:
```diff
@@ -0,0 +1,11 @@
+.parent {
+	display: grid;
+	grid-template-columns: repeat(5, 1fr);
+	grid-template-rows: repeat(5, 1fr);
+	grid-column-gap: 0px;
+	grid-row-gap: 0px;
+}
```
2 spaces for indentation
```suggestion
.parent {
  display: grid;
  grid-template-columns: repeat(5, 1fr);
  grid-template-rows: repeat(5, 1fr);
  grid-column-gap: 0px;
  grid-row-gap: 0px;
}
```

lra-wtag replied:
please resolve this

meb-wtag committed:
Update grid.css

evk-wtag commented on spec/models/journal_entry_spec.rb:
```diff
@@ -1,5 +1,7 @@
 require 'rails_helper'
 RSpec.describe JournalEntry, type: :model do
+  let!(:user) { FactoryBot.create(:user) }
```
Please study the difference between let and let!
```suggestion
  let(:user) { FactoryBot.create(:user) }
```

evk-wtag commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,15 @@
+  end
+
+  def require_login
+    return if current_user == @user
```
Please add a new line after guard clause

evk-wtag commented:
Let's follow this for every PR:
- check if all files have a new line at the end of files
- add i18n instead of using hard coded texts

evk-wtag commented on app/assets/.DS_Store:
Let's not push .DS_Store files

lra-wtag replied:
lets revert this file

meb-wtag committed:
Changes 3

meb-wtag committed:
-

lra-wtag commented on app/models/user.rb:
```diff
@@ -0,0 +1,11 @@
+  has_one_attached :profile_picture
+  has_secure_password :password, validations: true
+
+  validates :username, presence: true, uniqueness: true
```
this user name is the email ?

meb-wtag replied:
until now no, it's a custom username

lra-wtag commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,95 @@
+                    <i class="fa fa-camera fa-2x"></i>
+                  </div>
+                  <div class="text-uppercase">
+                    Update <br /> Profile Photo
```
i18n please

lra-wtag commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,95 @@
+          <p class="text-muted mb-1"><%= @user.description %></p>
+          <p class="text-muted mb-4"><%= @user.country %> <%= @user.city %></p>
+          <div class="d-flex justify-content-center mb-2">
+            <%= link_to 'Edit Profile', edit_user_path(@user), class: "btn btn-primary"%>
```
i18n please

lra-wtag commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,95 @@
+          <div class="d-flex justify-content-center mb-2">
+          <%= link_to "Log out", logout_path, class: "btn btn-danger", data: {
+                  turbo_method: :delete,
+                  turbo_confirm: "Are you sure?"
```
i18n please

lra-wtag replied:
i18n is needed here

lra-wtag commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,95 @@
+              <p class="mb-0"><%= t('user.show.adress') %></p>
+            </div>
+            <div class="col-sm-9">
+              <p class="text-muted mb-0"><%= @user.country %>, adress...</p>
```
```suggestion
              <p class="text-muted mb-0"><%= @user.country %></p>
```

lra-wtag commented on config/webpack/environment.js:
```diff
@@ -0,0 +1,13 @@
+  })
+);
+
+module.exports = environment;
```
please add an extra line at the end of the file

lra-wtag reviewed (changes requested):
Moving forward lets create smaller PRs and not have everything in one PR. It becomes difficult to review

meb-wtag committed:
Deleted .DS_Store files

lra-wtag commented on config/locales/en.yml:
```diff
@@ -40,3 +41,57 @@ en:
+    edit: Edit Profile
+  profile_picture:
+    update: Update
+    profile_picture: Profile Picture
```
please add an extra line at the end of the file

meb-wtag committed:
Update en.yml

meb-wtag committed:
All changes made

Ready to merge
